zinc-k8s-operator/src/charm.py

```diff
@@ -0,0 +1,145 @@
+        framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
+        framework.observe(self.on["zinc"].pebble_ready, self._on_zinc_pebble_ready)
+        framework.observe(self.on.get_admin_password_action, self._on_get_admin_password)
+
+    def _on_collect_unit_status(self, event: ops.CollectStatusEvent):
+        """Report the status of the unit."""
+        event.add_status(self._get_unit_status_for_container())
+        # For a more complex charm, there might be other aspects to check.
+
+    def _get_unit_status_for_container(self) -> ops.StatusBase:
+        """Determine the status of the workload container and return an appropriate unit status."""
+        try:
+            service_info = self._pebble.get_service("zinc")
+        except (ops.pebble.APIError, ops.ModelError):
+            return ops.MaintenanceStatus("waiting for Pebble to be ready")
+        if not service_info.is_running():
+            return ops.MaintenanceStatus("waiting for Zinc to start")
+        # Zinc didn't exit on startup, but it might not be ready yet, so check its health.
+        checks = self._pebble.get_checks(level=ops.pebble.CheckLevel.READY)
+        for check_info in checks.values():
+            if check_info.status == ops.pebble.CheckStatus.DOWN:
+                return ops.MaintenanceStatus("waiting for Zinc to be ready")
+                # TODO: Should we return BlockedStatus instead?
+                #       To inspect failing checks, the admin could run:
+                #       ```
+                #       juju ssh <unit> PEBBLE_SOCKET=/charm/containers/zinc/pebble.socket \
+                #       /charm/bin/pebble checks --level ready
+                #       ```
+                #       The command needs to be like this because the OCI image for Zinc has
+                #       no shell binaries (so `juju ssh --container zinc <unit> ...` won't work).
+        return ops.ActiveStatus()
+
+    def _on_zinc_pebble_ready(self, event: ops.PebbleReadyEvent):
+        """Use Pebble to start Zinc."""
+        self.unit.status = ops.MaintenanceStatus("starting Zinc")
+        if not self._update_zinc_config_from_peer_relation():
+            if self.unit.is_leader():
+                self._update_peer_relation_from_zinc_config()
+            else:
+                event.defer() # Wait for the leader to put Zinc config in the peer relation
+                return
+        self._add_pebble_layer()
+        self._pebble.replan()
+        version = self._zinc.get_version() # Retries with delay (in case of slow startup)
+        self.unit.set_workload_version(version)
+
+    def _add_pebble_layer(self):
+        # The OCI image for Zinc only has two binaries: zincsearch and go-runner.
+        # We use go-runner to achieve the equivalent of `bash -c '/bin/zincsearch | tee PATH'`.
+        command = "/bin/go-runner --log-file=/var/lib/zincsearch/zinc.log --also-stdout=true --redirect-stderr=true /bin/zincsearch"
```

Comment:
Is the go-runner tool just used for redirecting stdout and stderr to the log file? Could we just run the zincsearch binary directly and let them use Pebble logging? (We wouldn't even need to use log forwarding, just the Pebble log buffer.) That way we could also have a slightly smaller OCI image.